core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -249,7 +249,14 @@ private[spark] class ExecutorAllocationManager(
    * yarn-client mode when AM re-registers after a failure.
    */
   def reset(): Unit = synchronized {
-    initializing = true
+    /**
+     * When some tasks need to be scheduled and initial executor = 0, resetting the initializing
+     * field may cause it to not be set to false in yarn.
+     * SPARK-20079: https://issues.apache.org/jira/browse/SPARK-20079
+     */
+    if (maxNumExecutorsNeeded() == 0) {
+      initializing = true
```

Comment:
> when a new AM registers the driver needs to tell it how many executors it wants.

When a AM registers, it leverages configuration to decide the initial number of executors should be created, not driver who told him how many executors it wants. That's why in the driver side if we don't change the executor number to match the AM side, we will meet the problem as mentioned above (because driver hasn't yet told AM the executor number).